kafka-ui-api/src/test/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDeTest.java

```diff
@@ -0,0 +1,183 @@
+      registryClient,
+      new BufSchemaRegistryClient()
+  );
+
+  @Nested
+  class Deserialize {
+
+    // copied from SchemaRegistryAwareRecordSerDe
+    @Test
+    void callsSchemaFormatterWhenValueHasMagicByteAndValidSchemaId() throws Exception {
+      AvroSchema schema = new AvroSchema(
+          "{"
+              + "  \"type\": \"record\","
+              + "  \"name\": \"TestAvroRecord1\","
+              + "  \"fields\": ["
+              + "    {"
+              + "      \"name\": \"field1\","
+              + "      \"type\": \"string\""
+              + "    },"
+              + "    {"
+              + "      \"name\": \"field2\","
+              + "      \"type\": \"int\""
+              + "    }"
+              + "  ]"
+              + "}"
+      );
+
+      String jsonValueForSchema = "{ \"field1\":\"testStr\", \"field2\": 123 }";
+
+      int schemaId = 1234;
+      when(registryClient.getSchemaById(schemaId)).thenReturn(schema);
+
+      var result = serde.deserialize(
+          new ConsumerRecord<>(
+              "test-topic",
+              1,
+              100,
+              Bytes.wrap("key".getBytes()),
+              bytesWithMagicByteAndSchemaId(schemaId, jsonToAvro(jsonValueForSchema, schema))
+          )
+      );
+
+      // called twice: once by serde code, once by formatter (will be cached)
+      verify(registryClient, times(2)).getSchemaById(schemaId);
+
+      assertThat(result.getKeySchemaId()).isNull();
+      assertThat(result.getKeyFormat()).isEqualTo(MessageFormat.UNKNOWN);
+      assertThat(result.getKey()).isEqualTo("key");
+
+      assertThat(result.getValueSchemaId()).isEqualTo(schemaId + "");
+      assertThat(result.getValueFormat()).isEqualTo(MessageFormat.AVRO);
+      assertJsonsEqual(jsonValueForSchema, result.getValue());
+    }
+
+    @Test
+    void callsBufFormatterWhenValueHasCorrectHeader() throws Exception {
+      ConsumerRecord<Bytes, Bytes> record =  new ConsumerRecord<>(
+          "test-topic",
+          1,
+          100,
+          Bytes.wrap("key".getBytes()),
+          Bytes.wrap("value".getBytes())
+      );
+      record.headers().add("PROTOBUF_TYPE", "protobuf_type".getBytes());
+      var result = serde.deserialize(record);
+
+      // verify schema registry was skipped
+      verify(registryClient, times(0)).getSchemaById(0);
+
+      assertThat(result.getKeySchemaId()).isNull();
+      assertThat(result.getKeyFormat()).isEqualTo(MessageFormat.UNKNOWN);
+      assertThat(result.getKey()).isEqualTo("key");
+
+      assertThat(result.getValueFormat()).isEqualTo(MessageFormat.UNKNOWN);
+    }
+
+    @Test
+    void callsBufFormatterWhenTopicIsCorrect() throws Exception {
+      ConsumerRecord<Bytes, Bytes> record =  new ConsumerRecord<>(
+          "test-topic.proto.foo",
+          1,
+          100,
+          Bytes.wrap("key".getBytes()),
+          Bytes.wrap("value".getBytes())
+      );
+      record.headers().add("PROTOBUF_TYPE", "protobuf_type".getBytes());
```

Comment:
I think we should remove this line so it tests that Buf is used if the topic is set even though the header isn't set

Reply:
good catch! Fixed